File: extension/core/src/main/java/io/holunda/camunda/bpm/data/writer/TaskServiceVariableWriter.java
package io.holunda.camunda.bpm.data.writer;

import io.holunda.camunda.bpm.data.factory.VariableFactory;
import org.camunda.bpm.engine.TaskService;
import org.camunda.bpm.engine.variable.VariableMap;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

/**
 * User task builder allowing for fluent variable setting.
 */
public class TaskServiceVariableWriter implements VariableWriter<TaskServiceVariableWriter> {

  private final TaskService taskService;
  private final String taskId;

  /**
   * Creates a builder working on a user task.
   */
  public TaskServiceVariableWriter(TaskService taskService, String taskId) {
    this.taskService = taskService;
    this.taskId = taskId;
  }

  @Override
  @NotNull
  public VariableMap variables() {
    return this.taskService.getVariablesTyped(this.taskId);
  }

  @Override
  @NotNull
  public VariableMap variablesLocal() {
    return this.taskService.getVariablesLocalTyped(this.taskId);
  }

  @Override
  @NotNull
  public <T> TaskServiceVariableWriter set(VariableFactory<T> factory, T value) {
    return this.set(factory, value, false);
  }

  @Override
  @NotNull
  public <T> TaskServiceVariableWriter set(VariableFactory<T> factory, T value, boolean isTransient) {
    factory.on(this.taskService, this.taskId).set(value, isTransient);
    return this;
  }

  @Override
  @NotNull
  public <T> TaskServiceVariableWriter setLocal(VariableFactory<T> factory, T value) {
    return this.setLocal(factory, value, false);
  }

  @Override
  @NotNull
  public <T> TaskServiceVariableWriter setLocal(VariableFactory<T> factory, T value, boolean isTransient) {
    factory.on(this.taskService, this.taskId).setLocal(value, isTransient);
    return this;
  }

  @Override
  @NotNull
  public <T> TaskServiceVariableWriter remove(VariableFactory<T> factory) {
    factory.on(this.taskService, this.taskId).remove();
    return this;
  }

  @Override
  @NotNull
  public <T> TaskServiceVariableWriter removeLocal(VariableFactory<T> factory) {
    factory.on(this.taskService, this.taskId).removeLocal();
    return this;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    TaskServiceVariableWriter that = (TaskServiceVariableWriter) o;

    if (!Objects.equals(taskService, that.taskService)) return false;
    return Objects.equals(taskId, that.taskId);
  }

  @Override
  public int hashCode() {
    int result = taskService != null ? taskService.hashCode() : 0;
    result = 31 * result + (taskId != null ? taskId.hashCode() : 0);
    return result;
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/writer/TaskServiceVariableWriter.kt
package io.holunda.camunda.bpm.data.writer

import io.holunda.camunda.bpm.data.factory.VariableFactory
import org.camunda.bpm.engine.TaskService
import org.camunda.bpm.engine.variable.VariableMap
import java.util.*
import java.util.function.Function

/**
 * User task builder allowing for fluent variable setting.
 */
class TaskServiceVariableWriter(private val taskService: TaskService, private val taskId: String) :
  VariableWriter<TaskServiceVariableWriter> {
  override fun variables(): VariableMap {
    return taskService.getVariablesTyped(taskId)
  }

  override fun variablesLocal(): VariableMap {
    return taskService.getVariablesLocalTyped(taskId)
  }

  override fun <T> set(variableFactory: VariableFactory<T>, value: T): TaskServiceVariableWriter {
    return this.set(variableFactory, value, false)
  }

  override fun <T> set(variableFactory: VariableFactory<T>, value: T, isTransient: Boolean): TaskServiceVariableWriter {
    variableFactory.on(taskService, taskId)[value] = isTransient
    return this
  }

  override fun <T> setLocal(variableFactory: VariableFactory<T>, value: T): TaskServiceVariableWriter {
    return this.setLocal(variableFactory, value, false)
  }

  override fun <T> setLocal(variableFactory: VariableFactory<T>, value: T, isTransient: Boolean): TaskServiceVariableWriter {
    variableFactory.on(taskService, taskId).setLocal(value, isTransient)
    return this
  }

  override fun <T> remove(variableFactory: VariableFactory<T>): TaskServiceVariableWriter {
    variableFactory.on(taskService, taskId).remove()
    return this
  }

  override fun <T> removeLocal(variableFactory: VariableFactory<T>): TaskServiceVariableWriter {
    variableFactory.on(taskService, taskId).removeLocal()
    return this
  }

  override fun <T> update(variableFactory: VariableFactory<T>, valueProcessor: Function<T, T>): TaskServiceVariableWriter {
    variableFactory.on(taskService, taskId).update(valueProcessor)
    return this
  }

  override fun <T> update(
    variableFactory: VariableFactory<T>,
    valueProcessor: Function<T, T>,
    isTransient: Boolean
  ): TaskServiceVariableWriter {
    variableFactory.on(taskService, taskId).update(valueProcessor, isTransient)
    return this
  }

  override fun <T> updateLocal(
    variableFactory: VariableFactory<T>,
    valueProcessor: Function<T, T>
  ): TaskServiceVariableWriter {
    variableFactory.on(taskService, taskId).updateLocal(valueProcessor)
    return this
  }

  override fun <T> updateLocal(
    variableFactory: VariableFactory<T>,
    valueProcessor: Function<T, T>,
    isTransient: Boolean
  ): TaskServiceVariableWriter {
    variableFactory.on(taskService, taskId).updateLocal(valueProcessor, isTransient)
    return this
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val that = other as TaskServiceVariableWriter
    return if (taskService != that.taskService) false else taskId == that.taskId
  }

  override fun hashCode(): Int {
    return Objects.hash(taskService, taskId)
  }
}
